File: src/main/java/org/zenbot/timetableupdater/domain/BusRoute.java
package org.zenbot.timetableupdater.domain;

import lombok.Data;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Data
@Document
public class BusRoute {

    @Id
    private String id;
    private String routename;
    private List<BusRouteLine> busRouteLines;

    public BusRouteLine getRoutePathByStartStopName(String startBusStopName) {
        Optional<BusRouteLine> routeLine = busRouteLines.stream()
                .filter(line -> line.getStartBusStop().equals(startBusStopName))
                .findFirst();
        if (routeLine.isPresent()) {
            return routeLine.get();
        } else {
            BusRouteLine result = new BusRouteLine();
            result.setBusStops(new ArrayList<>());
            return result;
        }
    }

    public boolean hasNoRoutePath(BusRouteLine busRouteLine) {
        return !busRouteLines.stream()
                .map(line -> line.getStartBusStop())
                .collect(Collectors.toList())
                .contains(busRouteLine.getStartBusStop());
    }
}

File: src/main/kotlin/org/zenbot/timetableupdater/domain/BusRoute.kt
package org.zenbot.timetableupdater.domain

import org.springframework.data.annotation.Id
import java.util.ArrayList
import java.util.stream.Collectors

data class BusRoute (
        @Id var id: String? = null,
        var routename: String = "",
        var busRouteLines: MutableList<BusRouteLine> = mutableListOf()
) {
    fun getRoutePathByStartStopName(startBusStopName: String) : BusRouteLine {
        val routeLine = busRouteLines.stream()
                .filter { line -> line.startBusStop == startBusStopName }
                .findFirst()
        return if (routeLine.isPresent) {
            routeLine.get()
        } else {
            val result = BusRouteLine()
            result.busStops = mutableListOf()
            result
        }
    }

    fun hasNoRoutePath(busRouteLine: BusRouteLine): Boolean {
        return !busRouteLines.stream()
                .map { line -> line.startBusStop }
                .collect(Collectors.toList())
                .contains(busRouteLine.startBusStop)
    }
}